File: src/java-sample/com/xtaudio/xt/RenderSimple.java
package com.xtaudio.xt;

public class RenderSimple {

    static double phase = 0.0;
    static final double FREQUENCY = 660.0;
    static final XtFormat FORMAT = new XtFormat(new XtMix(44100, XtSample.FLOAT32), 0, 0, 1, 0);

    static void render(XtStream stream, Object input, Object output, int frames,
            double time, long position, boolean timeValid, long error, Object user) {

        for (int f = 0; f < frames; f++) {
            phase += FREQUENCY / FORMAT.mix.rate;
            if (phase >= 1.0)
                phase = -1.0;
            ((float[]) output)[f] = (float) Math.sin(phase * Math.PI);
        }
    }

    public static void main(String[] args) throws Exception {

        try (XtAudio audio = new XtAudio(null, null, null, null)) {
            XtService service = XtAudio.getServiceBySetup(XtSetup.CONSUMER_AUDIO);
            try (XtDevice device = service.openDefaultDevice(true)) {
                if (device != null && device.supportsFormat(FORMAT)) {

                    XtBuffer buffer = device.getBuffer(FORMAT);
                    try (XtStream stream = device.openStream(FORMAT, true, false,
                            buffer.current, RenderSimple::render, null, null)) {
                        stream.start();
                        Thread.sleep(1000);
                        stream.stop();
                    }
                }
            }
        }
    }
}

File: src/cpp/sample/RenderSimple.cpp
#define _USE_MATH_DEFINES 1
#include <xt/cpp/XtCpp.hpp>

#include <cmath>
#include <chrono>
#include <thread>
#include <cstdint>

static float _phase = 0.0f;
static const float Frequency = 440.0f;
static const Xt::Channels Channels(0, 0, 1, 0);
static const Xt::Mix Mix(44100, Xt::Sample::Float32);
static const Xt::Format Format(Mix, Channels);

static float NextSample()
{
  _phase += Frequency / Mix.rate;
  if (_phase >= 1.0f) _phase = -1.0f;
  return sinf(2.0f * _phase * static_cast<float>(M_PI));
}

static void OnBuffer(const Xt::Stream& stream, const Xt::Buffer& buffer, void* user)
{
  float* output = static_cast<float*>(buffer.output);
  for (int32_t f = 0; f < buffer.frames; f++) output[f] = NextSample();
}

int RenderSimpleMain()
{
  std::unique_ptr<Xt::Platform> platform = Xt::Audio::Init("", nullptr, nullptr);
  Xt::System system = Xt::Audio::SetupToSystem(Xt::Setup::ConsumerAudio);
  std::unique_ptr<Xt::Service> service = platform->GetService(system);
  if (!service) return 0;

  std::unique_ptr<Xt::Device> device = service->OpenDefaultDevice(true);
  if (!device || !device->SupportsFormat(Format)) return 0;

  double bufferSize = device->GetBufferSize(Format).current;
  Xt::StreamParams streamParams(true, OnBuffer, nullptr);
  Xt::DeviceStreamParams deviceParams(streamParams, Format, bufferSize);
  std::unique_ptr<Xt::Stream> stream = device->OpenStream(deviceParams, nullptr);
  stream->Start();
  std::this_thread::sleep_for(std::chrono::seconds(2));
  stream->Stop();

  return 0;
}